### PR TITLE
Auto-Encode `values` Provided an Instance of `Schema[A]` Exists

### PR DESCRIPTION
**Pull request by ASRagab**
Closes zio/zio-jdbc#10

One issue with my approach, is that it still requires the explicitly declared implicit to be in scope 

```scala
implicit val userEncoder: JdbcEncoder[User] = JdbcEncoder.fromSchema
```

I ran into a problem in overloading the `.values` call and summoning the encoder from the schema there, since the overload would have been ambiguous (different only in the type of implicit parameter available).

The first solution would be to name the `.values` method for schema encoded values something different, but that did not feel very ergonomic or valuable.

The second approach I tried is to change the implicit parameter such that it could take a Schema or JdbcEncoder, i.e something like this:

```scala
  object implicits {
    case class EncoderOrSchema[A](e: Either[JdbcEncoder[A], Schema[A]])
    case class DecoderOrSchema[A](e: Either[JdbcDecoder[A], Schema[A]])

    object EncoderOrSchema {
      implicit def fromEncoder[A](implicit encoder: JdbcEncoder[A]): EncoderOrSchema[A] = EncoderOrSchema(Left(encoder))
      implicit def fromSchema[A](implicit schema: Schema[A]): EncoderOrSchema[A]        = EncoderOrSchema(Right(schema))
    }

    object DecoderOrSchema {
      implicit def fromDecoder[A](implicit decoder: JdbcDecoder[A]): DecoderOrSchema[A] = DecoderOrSchema(Left(decoder))
      implicit def fromSchema[A](implicit schema: Schema[A]): DecoderOrSchema[A]        = DecoderOrSchema(Right(schema))
    }
  }
```

and then have the `.values` call pattern match on the implicit that was available:

```scala
  def values[B](
    b: B,
    bs: B*
  )(implicit encoderOrSchema: EncoderOrSchema[B], ev: A <:< ZResultSet): SqlFragment =
    encoderOrSchema.e match {
      case Right(schema) => values(b +: bs)(JdbcEncoder.fromSchema(schema), ev)
      case Left(encoder) => values(b +: bs)(encoder, ev)
    }
````
    
This sort of worked, but for tuples it was broken since it seems there is a schema and jdbc encoder always in scope, likely due to the requisite imports. It also misbehaved in 2.12 with another compiler exception.
 
Maybe there's a better standard solution out there for this kind of problem, would enjoy other's input.

### CODE REVIEW

File: core/src/main/scala/zio/jdbc/Sql.scala [hunk 1]
```diff
@@ -71,14 +71,14 @@ final class Sql[+A](
...
-    val paramsString = if (params.length > 0) {
+    val paramsString = if (params.nonEmpty) {
       ", " + params.mkString(", ")
     } else ""
 
     s"Sql(${sql.result()}${paramsString})"
   }
 
-  def values[B](
+  private def values[B](
```
Comment (jdegoes) [inline core/src/main/scala/zio/jdbc/Sql.scala:81]: Was it intentional to make this method `private`?

Comment (ASRagab) [reply core/src/main/scala/zio/jdbc/Sql.scala:81]: @jdegoes it was but since the last PR you just merged, it is no longer necessary will change.

**Comment by jdegoes** [comment]
I think we do need to solve the implicit problem. I have been punting on that, until the library is a little further along, because it's not a critical issue but seems to be tricky.

**Comment by jdegoes** [comment]
Thank you for this contribution! 🙏

**Pull request merged**